base/loading.jl

```diff
@@ -315,25 +316,66 @@ all platforms, including those with case-insensitive filesystems like macOS and
 Windows.
 """
 function require(mod::Symbol)
-    _require(mod)
+    existed = root_module_exists(mod)
+    M = _require(mod)
     # After successfully loading, notify downstream consumers
     if toplevel_load[] && myid() == 1 && nprocs() > 1
         # broadcast top-level import/using from node 1 (only)
         @sync for p in procs()
             p == 1 && continue
             @async remotecall_wait(p) do
-                if !isbindingresolved(Main, mod) || !isdefined(Main, mod)
-                    _require(mod)
-                end
+                _require(mod)
+                nothing
             end
         end
     end
-    for callback in package_callbacks
-        invokelatest(callback, mod)
+    if !existed
+        for callback in package_callbacks
+            invokelatest(callback, mod)
+        end
+    end
+    return M
+end
+
+const loaded_modules = ObjectIdDict()
+const module_keys = ObjectIdDict()
+
+function register_root_module(key, m::Module)
```

Comment:
This seems like a bad idea. Maybe `require` should set a global flag (`haskey(package_locks, m)` or perhaps better would be `isempty(package_locks)`) that disables rooting of the object in `Main.m`?

Reply:
This function is really only meant for internal use. Nobody but the loading code should call it. Does that address your concern, or is there something else?

Reply:
This design appears to assume there will be a 1-to-1 correlation between `require` and module definition, which doesn't always happen now. I think a simpler design would be handle that decision point during construction, rather than post-factum mutation.

Reply:
I think we'll be able to handle it either way. If loading something defines 0 modules, we can store `nothing` here instead of a module (to provide "load once" semantics). If a file defines multiple modules, you won't be able to reference all of them without a distinct `using` statement for each one you want anyway. But we should definitely discourage or disallow that. @StefanKarpinski even briefly floated the idea of removing the `module` keyword (or making it optional), and having the loader make modules instead (thus enforcing 1-to-1 file-to-module by construction).